go/vt/vtgate/engine/fk_cascade.go

```diff
@@ -0,0 +1,180 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+type FkChild struct {
+	BVName string
+	Cols   []int // indexes
+	Exec   Primitive
+}
+
+// FkCascade is a primitive that implements foreign key cascading using Selection as values required to execute the FkChild Primitives.
+// On success, it executes the Parent Primitive.
+type FkCascade struct {
+	Selection Primitive
+	Children  []FkChild
+	Parent    Primitive
+
+	txNeeded
+}
+
+// RouteType implements the Primitive interface.
+func (fkc *FkCascade) RouteType() string {
+	return "FkCascade"
+}
+
+// GetKeyspaceName implements the Primitive interface.
+func (fkc *FkCascade) GetKeyspaceName() string {
+	return fkc.Parent.GetKeyspaceName()
+}
+
+// GetTableName implements the Primitive interface.
+func (fkc *FkCascade) GetTableName() string {
+	return fkc.Parent.GetTableName()
+}
+
+// GetFields implements the Primitive interface.
+func (fkc *FkCascade) GetFields(_ context.Context, _ VCursor, _ map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "[BUG] GetFields should not be called")
+}
+
+// TryExecute implements the Primitive interface.
+func (fkc *FkCascade) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	// Execute the Selection primitive to find the rows that are going to modified.
+	// This will be used to find the rows that need modification on the children.
+	selectionRes, err := vcursor.ExecutePrimitive(ctx, fkc.Selection, bindVars, wantfields)
+	if err != nil {
+		return nil, err
+	}
+
+	// If no rows are to be modified, there is nothing to do.
+	if len(selectionRes.Rows) == 0 {
+		return &sqltypes.Result{}, nil
+	}
+
+	for _, child := range fkc.Children {
+		// We create a bindVariable for each Child
+		// that stores the tuple of columns involved in the fk constraint.
+		bv := &querypb.BindVariable{
+			Type: querypb.Type_TUPLE,
+		}
+		for _, row := range selectionRes.Rows {
+			// Create a tuple from each Row.
+			tuple := &querypb.Value{
+				Type: querypb.Type_TUPLE,
+			}
+			for _, colIdx := range child.Cols {
+				tuple.Values = append(tuple.Values,
+					sqltypes.ValueToProto(row[colIdx]))
+			}
+			bv.Values = append(bv.Values, tuple)
+		}
+		// Execute the child primitive, and bail out incase of failure.
+		// Since this Primitive is always executed in a transaction, the changes should
+		// be rolled back incase of an error.
+		bindVars[child.BVName] = bv
```

Comment:
did you consider cloning the bindVars map to separate between the one sent to the cascades and the one sent to the Selection and the Parent?

Reply:
we can do it when we see the need for it, after using it in the plan and checking on what additional information is sent down which can be avoided.